torch/distributions/transformed_distribution.py

```diff
@@ -32,11 +32,9 @@ def __init__(self, base_distribution, transforms):
         event_dim = max([len(self.base_dist.event_shape)] + [t.event_dim for t in self.transforms])
         batch_shape = shape[:len(shape) - event_dim]
         event_shape = shape[len(shape) - event_dim:]
-        super(TransformedDistribution, self).__init__(batch_shape, event_shape)
+        super(TransformedDistribution, self).__init__(batch_shape, event_shape, validate_args=validate_args)
 
-    @constraints.dependent_property
-    def params(self):
-        return self.base_dist.params  # TODO add params of transforms?
+    params = {}
```

Comment:
Why do we have this in here?

Reply:
@lazypanda1 is fixing a type error here (my fault). `.params` should a dict mapping strings to `Constraint` objects, where the strings are kwargs keys of the distribution's `__init__()` method. However none of the arguments to `TransformedDistribution.__init__()` are tensors (the first arg is a `Distribution` and the second is a list of `Transform`s), hence none of the args can be constrained by `Constraint` objects, and hence `.params` should be empty.

It's fine for subclasses of `TransformedDistribution` to define a tensor-parameterized `__init__()` method and corresponding entries in `.params`.

Reply:
Yeah, but why can't we just return the parameters of the base distribution?

Reply:
@apaszke No, that does work as then it starts looking for the subclass parameters in the `TransformedDistribution` class which fails.

Reply:
Sorry, I didn't get that. What starts looking for the parameters?

Reply:
`TestValidation.test_valid` test fails with the following error:


```
self = <test_distributions.TestValidation testMethod=test_valid>

    def test_valid(self):
        for Dist, params in EXAMPLES:
            for i, param in enumerate(params):
>               Dist(validate_args=True, **param)

test/test_distributions.py:3065: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
torch/distributions/transformed_distribution.py:35: in __init__
    super(TransformedDistribution, self).__init__(batch_shape, event_shape, validate_args=validate_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = TransformedDistribution(), batch_shape = torch.Size([2, 3]), event_shape = torch.Size([]), validate_args = True

    def __init__(self, batch_shape=torch.Size(), event_shape=torch.Size(), validate_args=None):
        self._batch_shape = batch_shape
        self._event_shape = event_shape
        if validate_args is not None:
            self._validate_args = validate_args
        if self._validate_args:
            if not constraints.is_dependent(self.params):
                for param, constraint in self.params.items():
                    if not constraints.is_dependent(constraint):
>                       if not constraint.check(self.__getattribute__(param)).all():
E                       AttributeError: 'TransformedDistribution' object has no attribute 'loc'

torch/distributions/distribution.py:33: AttributeError

```

This is what i would have in the code:
```
class TransformedDistribution(Distribution):
  ....

    @constraints.dependent_property
    def params(self):
        return self.base_dist.params
```

What do u suggest?

Reply:
@apaszke To support generic programming, the `d.param` should refer only to args of that `d.__init__`. If you want to examine the parameters of the base distribution of a transformed distribution, then you should call `my_transformed_dist.base_dist.params`. Note that most instances of transformed by not only their `base_dist` parameters but also parameters of the transforms, and there is no simple way to gather all those parameters automatically.

Reply:
Hey guys, do the changes look fine? Please let me know

Reply:
So it's not so much about having the parameters be the `__init__` arguments, but it's more about them being attributes of the distributed later, correct? We could add `__getattr__` for `TransformedDistribution` such that it allows access to `base_dist` params. Alternatively we can do this, but it's a bit weird that we're lying about the parameter set.

Reply:
@apaszke What do you think of declaring the non-tensor args as `dependent`?
```py
params = {'base_dist': constraints.dependent, 'transforms': constraints.dependent}
```

Reply:
@fritzo Yes. That passes the tests!

Reply:
hmm maybe I misunderstand what `params` are supposed to be. I thought it should be a mapping from names of attributes (that are tensors) to constraints imposed on them. This, however, is completely different. @fritzo what are the semantics that you have in mind?

Reply:
NB I can merge it if it's blocking for you somehow, but I wanted to clarify that. Right now the API seems inconsistent

Reply:
@fritzo let me know if you would like me to push this change. I want to merge this today.

Reply:
@apaszke maybe I also misunderstand, sorry for any confusion. To avoid blocking @lazypanda1 why don't we try to revert this one change and you and I can discuss on Slack or Zoom or sth. 

@lazypanda1 apologies for confusion. Does it work if you revert this change and simply add a guard to the tests
```py
if constraints.is_dependent(my_class.params):
    continue
```

Reply:
Done.